pkg/skaffold/initializer/init.go

```diff
@@ -538,40 +542,80 @@ func printAnalyzeJSON(out io.Writer, skipBuild bool, pairs []builderImagePair, u
 	return err
 }
 
-func walk(dir string, force, enableJibInit bool, validateBuildFile func(bool, string) ([]InitBuilder, error)) ([]string, []InitBuilder, error) {
+func walk(dir string, force, enableJibInit bool) ([]string, []InitBuilder, error) {
 	var potentialConfigs []string
 	var foundBuilders []InitBuilder
-	err := filepath.Walk(dir, func(path string, f os.FileInfo, e error) error {
-		if f.IsDir() && util.IsHiddenDir(f.Name()) {
-			logrus.Debugf("skip walking hidden dir %s", f.Name())
-			return filepath.SkipDir
-		}
-		if f.IsDir() || util.IsHiddenFile(f.Name()) {
-			return nil
+
+	var dirCallback func(path string, findBuilders bool) error
+	dirCallback = func(path string, findBuilders bool) error {
+		dirents, err := godirwalk.ReadDirents(path, nil)
+		if err != nil {
+			return err
 		}
-		if IsSkaffoldConfig(path) {
-			if !force {
-				return fmt.Errorf("pre-existing %s found", path)
+
+		var directories []*godirwalk.Dirent
+		findBuildersInDirectories := true
+		sort.Sort(dirents)
+
+		// Traverse files
+		for _, file := range dirents {
+			if util.IsHiddenFile(file.Name()) || util.IsHiddenDir(file.Name()) {
+				continue
+			}
+
+			// If we found a directory, keep track of it until we've gone through all the files first
+			if file.IsDir() {
+				directories = append(directories, file)
+				continue
+			}
+
+			filePath := filepath.Join(path, file.Name())
+			if findBuildersInDirectories, err = checkFile(filePath, force, findBuilders, enableJibInit, &potentialConfigs, &foundBuilders); err != nil {
+				return err
 			}
-			logrus.Debugf("%s is a valid skaffold configuration: continuing since --force=true", path)
-			return nil
-		}
-		if IsSupportedKubernetesFileExtension(path) {
-			potentialConfigs = append(potentialConfigs, path)
-			return nil
 		}
-		// try and parse build file
-		if builderConfigs, err := validateBuildFile(enableJibInit, path); builderConfigs != nil {
-			for _, buildConfig := range builderConfigs {
-				logrus.Infof("existing builder found: %s", buildConfig.Describe())
-				foundBuilders = append(foundBuilders, buildConfig)
+
+		// Traverse into subdirectories
+		for _, dir := range directories {
+			err = dirCallback(filepath.Join(path, dir.Name()), findBuildersInDirectories)
```

Comment:
Eventually, you end up traversing all the sub-directories that you think you should, so I think it doesn't matter if you do BFS or DFS (if I'm not mistaken about this). Then I think doing DFS would simplify the code, removing the `directories` array. No?

Reply:
I think it does matter that we do a BFS, because we want to search all the files in a directory before deciding whether to search subdirectories. For example, in a multi-module maven project, if we have the sub modules in subdirectories, we want to find the parent pom.xml before we enter those directories, or we risk running the `jib:_skaffoldInit` goal on a sub module instead of the parent.

Reply:
Oh, yeah, my misunderstanding.